yjit/src/codegen.rs

```diff
@@ -1329,6 +1329,28 @@ fn guard_object_is_array(
     asm.jne(side_exit.into());
 }
 
+// Guard that Opnd is a T_OBJECT.
+// Opnd must never be an immediate, so use this with guard_object_is_heap
+fn guard_object_is_object(
+    asm: &mut Assembler,
+    object_opnd: Opnd,
+    side_exit: CodePtr,
+) {
+    asm.comment("guard object is object");
+
+    // Pull out the type mask
+    let flags_opnd = Opnd::mem(
+        8 * SIZEOF_VALUE as u8,
+        object_opnd,
+        RUBY_OFFSET_RBASIC_FLAGS,
+    );
+    let flags_opnd = asm.and(flags_opnd, (RUBY_T_MASK as u64).into());
+
+    // Compare the result with T_OBJECT
+    asm.cmp(flags_opnd, (RUBY_T_OBJECT as u64).into());
+    asm.jne(side_exit.into());
```

Comment:
I would have expected the type in the context to be refined to `Object` after this type check, it feels like that should be part of this guard operation so that it happens automatically.

Maybe that should be done as part of a separate PR because I realize it's orthogonal to the change you're making here, but we do that in `guard_two_fixnums`, but we don't do it in `guard_object_is_array` for instance, which means we're not taking full advantage of the checks we're doing.

Reply:
TBH I didn't really want to write this guard 😅

If we integrate T_OBJECT in to the shape, then we don't need to have this guard.  I decided to see how hard it would be to add T_OBJECT and [it wasn't too bad](https://github.com/ruby/ruby/pull/6758).

Once I merge that, I'll just delete this code.